pandas/core/frame.py

```diff
@@ -11570,32 +11637,107 @@ def all(
             result = result.__finalize__(self, method="all")
         return result
 
+    # error: Signature of "min" incompatible with supertype "NDFrame"
+    @overload  # type: ignore[override]
+    def min(
+        self,
+        *,
+        axis: Axis = ...,
+        skipna: bool = ...,
+        numeric_only: bool = ...,
+        **kwargs,
+    ) -> Series:
+        ...
+
+    @overload
+    def min(
+        self,
+        *,
+        axis: None,
+        skipna: bool = ...,
+        numeric_only: bool = ...,
+        **kwargs,
+    ) -> Any:
+        ...
+
+    @overload
+    def min(
+        self,
+        *,
+        axis: Axis | None,
+        skipna: bool = ...,
+        numeric_only: bool = ...,
+        **kwargs,
+    ) -> Series | Any:
```

Comment:
Similar to the "Any trick" at typeshed https://github.com/python/typeshed/blob/main/CONTRIBUTING.md#the-any-trick

- calling code needs to be able to handle a Series
- Any because: we do not know the concrete return type when `axis=None`
- and Any because: type checkers should not warn about unused code (isinstance checks or other special casing in the calling code)

Reply:
> i'm not up to speed on the pandas current typing processes.. typing in pandas codebase is still for pandas code integrity checks only (i.e. not public) and all users/projects should use the stubs?

I believe that is the intention (I try to stay out of pandas politics, so I might not be up-to-date). I forgot whether it was a stated goal also shared by others: trickle annotations (when appropriate) from pandas-stubs back to pandas (or to directly add them to pandas).

I want to highlight that Pyright analyzes the in-line annotations of pandas (unless pandas-stubs is also installed). So it is definitely helpful to also have public API annotated in pandas.

(Pandas-stubs also makes a few "opinionated" choices, such as only allowing certain types in Series even though anything can go in them. So there might be a group of users who prefer the pandas annotations.)

> do we see that here?

Public API is also used by pandas internally (not sure about this particular method), so I don't see a reason not to add it (it also is a form of documentation). The other options: a) not having a return type (<-> disallow_incomplete_defs) or b) claiming that it returns only Series or only Any are, from my perspective, worse than the proposed change. What is your proposal?

While I like to avoid Any, I'm open to using it for pandas when literally anything can be returned (even when "opinionated" choices would allow non-Any return types).

Reply:
Yep, my intentions weren't clear. I have no objections to the type annotation here.

My question was more to clarify why you felt the need to comment here. i.e. were you expecting it to be challenged or is the annotation not in keeping with current practices?

Or is it that the pandas-stubs philosophy about stubs changed. I know that a long while back there were differences of opinion on whether stubs should be accurate or only cover the typical uses. When covering only typically uses then you could expect warnings about unused code paths.

So just to confirm, this matches the return type in the stubs? That's fine.

Reply:
I added the comments because we haven't used the "Any trick" in pandas before. At first sight, it might look weird: could  `Series | Any` be simplified to  `Any`?

Some of these methods are not yet annotated in pandas-stubs, other methods could have a more precise annotation in pandas-stubs for Series (as Series is generic).

I don't think there was a change in philosophy at pandas-stubs. Luckily, it seems to be accepted, basically no issues about being too strict (too typical).

Reply:
> I added the comments because we haven't used the "Any trick" in pandas before.

Looking through the code you will find some occurrences. IIRC this has been discussed in the past. And so I have no objections to this pattern because of the advantages over just using `Any`.

> I don't think there was a change in philosophy at pandas-stubs. Luckily, it seems to be accepted, basically no issues about being too strict (too typical).

Yes, didn't want to reopen the too narrow vs too wide argument. Warnings about unused code paths is a downside of the too narrow camp.

Reply:
Oh, yes - when grepping for it, I see a few cases. Didn't remember that.